File: simple_ga.hs
import System.Random
import qualified Data.List as List

target :: String
target = "Hello World"

-- Create a random individual
createIndividual :: StdGen -> (String, StdGen)
createIndividual gen =
  let (randomChars, newGen) = foldr generateRandomChar ([], gen) [1..length target]
  in (randomChars, newGen)
  where
    generateRandomChar :: a -> ([Char], StdGen) -> ([Char], StdGen)
    generateRandomChar _ (chars, g) =
      let (randInt, newGen) = randomR (0, charSetSize - 1) g :: (Int, StdGen)
      in (charSet !! randInt : chars, newGen)

-- Create a population of individuals
createPopulation :: Int -> StdGen -> ([String], StdGen)
createPopulation size gen =
  foldr generateIndividual ([], gen) [1..size]
  where
    generateIndividual :: a -> ([String], StdGen) -> ([String], StdGen)
    generateIndividual _ (population, g) =
      let (individual, newGen) = createIndividual g
      in (individual : population, newGen)

-- Calculate the fitness of an individual
fitnessFunction :: String -> Int
fitnessFunction individual =
  sum $ zipWith (\a b -> if a == b then 1 else 0) target individual

-- Select parents from the population
selectParents :: [String] -> StdGen -> ([String], StdGen)
selectParents population gen =
  let fitnesses = map fitnessFunction population
      totalFitness = sum fitnesses
      probabilities = map (\fitness -> fromIntegral fitness / fromIntegral totalFitness) fitnesses
      (index1, newGen1) = weightedRandomChoice probabilities gen
      (index2, newGen2) = weightedRandomChoice probabilities newGen1
  in ([population !! index1, population !! index2], newGen2)

-- Helper function for weighted random choice
weightedRandomChoice :: [Float] -> StdGen -> (Int, StdGen)
weightedRandomChoice weights gen =
  let (randVal, newGen) = random gen :: (Float, StdGen)
      cumulativeWeights = List.scanl (+) 0 weights
      index = findIndex randVal cumulativeWeights
  in (index, newGen)

-- Find the index where the cumulative weight exceeds the random value
findIndex :: Float -> [Float] -> Int
findIndex randVal cumulativeWeights =
  case dropWhile (<= randVal) cumulativeWeights of
    [] -> length cumulativeWeights - 1
    (x:_) -> List.findIndex (==x) cumulativeWeights

-- Perform crossover between two parents
crossover :: String -> String -> Float -> StdGen -> (String, StdGen)
crossover parent1 parent2 crossoverRate gen =
  if randVal < crossoverRate
  then
    let crossoverPoint = randomR (0, length target - 1) :: StdGen -> (Int, StdGen)
        (point, newGen) = crossoverPoint gen
    in (take point parent1 ++ drop point parent2, newGen)
  else
    (if randomBool gen then parent1 else parent2, gen)
  where
    (randVal, _) = random gen :: (Float, StdGen)

-- Helper function for random boolean value
randomBool :: StdGen -> Bool
randomBool gen =
  let (randInt, _) = randomR (0, 1) gen :: (Int, StdGen)
  in randInt == 1

-- Mutate an individual
mutate :: String -> Float -> StdGen -> (String, StdGen)
mutate individual mutationRate gen =
  foldr mutateChar ([], gen) [0..length individual - 1]
  where
    mutateChar :: Int -> ([Char], StdGen) -> ([Char], StdGen)
    mutateChar index (mutatedChars, g) =
      if randVal < mutationRate
      then
        let (randInt, newGen) = randomR (0, charSetSize - 1) g :: (Int, StdGen)
        in (charSet !! randInt : mutatedChars, newGen)
      else
        (individual !! index : mutatedChars, g)
      where
        (randVal, _) = random g :: (Float, StdGen)

-- Run the genetic algorithm
geneticAlgorithm :: Int -> Int -> Float -> Float -> StdGen -> IO ()
geneticAlgorithm populationSize generations mutationRate crossoverRate initialGen = do
  let (initialPopulation, _) = createPopulation populationSize initialGen
  evolve initialPopulation 0 initialGen
  where
    evolve :: [String] -> Int -> StdGen -> IO ()
    evolve population generation gen
      | generation >= generations = putStrLn "Target not found"
      | target `elem` population = putStrLn $ "Target found in generation " ++ show generation ++ ": " ++ target
      | otherwise = do
          let sortedPopulation = List.sortBy (compare `on` (negate . fitnessFunction)) population
          putStrLn $ "Generation " ++ show generation ++ ": " ++ head sortedPopulation
          let (parentPair, newGen1) = selectParents population gen
              [parent1, parent2] = parentPair
              (child, newGen2) = crossover parent1 parent2 crossoverRate newGen1
              (mutatedChild, newGen3) = mutate child mutationRate newGen2
              newPopulation = take (populationSize - 1) sortedPopulation ++ [mutatedChild]
          evolve newPopulation (generation + 1) newGen3

-- Charset and its size
charSet :: String
charSet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ "

charSetSize :: Int
charSetSize = length charSet

-- Comparison function
on :: (Ord b) => (a -> b) -> a -> a -> Ordering
on f x y = compare (f x) (f y)

main :: IO ()
main = do
  gen <- getStdGen
  let populationSize = 100
      generations = 1000
      mutationRate = 0.01
      crossoverRate = 0.7
  geneticAlgorithm populationSize generations mutationRate crossoverRate gen

File: simple_ga.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define TARGET "Hello World"
#define TARGET_LENGTH 11
#define POPULATION_SIZE 100
#define GENERATIONS 1000
#define MUTATION_RATE 0.01
#define CROSSOVER_RATE 0.7
#define CHARSET "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ "
#define CHARSET_SIZE 53

typedef struct {
    char genes[TARGET_LENGTH + 1];
    int fitness;
} Individual;

// Function prototypes
Individual create_individual();
void create_population(Individual population[]);
int fitness_function(Individual individual);
void select_parents(Individual population[], Individual *parent1, Individual *parent2);
Individual crossover(Individual parent1, Individual parent2);
void mutate(Individual *individual);
void genetic_algorithm();

int main() {
    // Seed the random number generator
    srand(time(NULL));

    genetic_algorithm();

    return 0;
}

Individual create_individual() {
    Individual individual;
    for (int i = 0; i < TARGET_LENGTH; i++) {
        individual.genes[i] = CHARSET[rand() % CHARSET_SIZE];
    }
    individual.genes[TARGET_LENGTH] = '\0'; // Null-terminate the string
    individual.fitness = 0; // Initialize fitness
    return individual;
}

void create_population(Individual population[]) {
    for (int i = 0; i < POPULATION_SIZE; i++) {
        population[i] = create_individual();
    }
}

int fitness_function(Individual individual) {
    int fitness = 0;
    for (int i = 0; i < TARGET_LENGTH; i++) {
        if (individual.genes[i] == TARGET[i]) {
            fitness++;
        }
    }
    return fitness;
}

void select_parents(Individual population[], Individual *parent1, Individual *parent2) {
    // Calculate fitness for each individual
    int fitnesses[POPULATION_SIZE];
    int total_fitness = 0;
    for (int i = 0; i < POPULATION_SIZE; i++) {
        fitnesses[i] = fitness_function(population[i]);
        total_fitness += fitnesses[i];
    }

    // Calculate selection probabilities
    float probabilities[POPULATION_SIZE];
    for (int i = 0; i < POPULATION_SIZE; i++) {
        probabilities[i] = (float)fitnesses[i] / total_fitness;
    }

    // Select two parents based on probabilities
    float rand1 = (float)rand() / RAND_MAX;
    float rand2 = (float)rand() / RAND_MAX;
    int index1 = 0, index2 = 0;
    float cumulative_probability = 0.0;

    for (int i = 0; i < POPULATION_SIZE; i++) {
        cumulative_probability += probabilities[i];
        if (rand1 < cumulative_probability) {
            index1 = i;
            break;
        }
    }

    cumulative_probability = 0.0;
    for (int i = 0; i < POPULATION_SIZE; i++) {
        cumulative_probability += probabilities[i];
        if (rand2 < cumulative_probability) {
            index2 = i;
            break;
        }
    }

    *parent1 = population[index1];
    *parent2 = population[index2];
}

Individual crossover(Individual parent1, Individual parent2) {
    Individual child;
    if ((float)rand() / RAND_MAX < CROSSOVER_RATE) {
        int crossover_point = rand() % TARGET_LENGTH;
        for (int i = 0; i < TARGET_LENGTH; i++) {
            if (i < crossover_point) {
                child.genes[i] = parent1.genes[i];
            } else {
                child.genes[i] = parent2.genes[i];
            }
        }
        child.genes[TARGET_LENGTH] = '\0';
    } else {
        // No crossover, just copy one of the parents
        if (rand() % 2 == 0) {
            strcpy(child.genes, parent1.genes);
        } else {
            strcpy(child.genes, parent2.genes);
        }
    }
    return child;
}

void mutate(Individual *individual) {
    for (int i = 0; i < TARGET_LENGTH; i++) {
        if ((float)rand() / RAND_MAX < MUTATION_RATE) {
            individual->genes[i] = CHARSET[rand() % CHARSET_SIZE];
        }
    }
}

void genetic_algorithm() {
    Individual population[POPULATION_SIZE];
    create_population(population);

    for (int generation = 0; generation < GENERATIONS; generation++) {
        // Sort the population (simple bubble sort for demonstration)
        for (int i = 0; i < POPULATION_SIZE - 1; i++) {
            for (int j = 0; j < POPULATION_SIZE - i - 1; j++) {
                if (fitness_function(population[j]) < fitness_function(population[j + 1])) {
                    Individual temp = population[j];
                    population[j] = population[j + 1];
                    population[j + 1] = temp;
                }
            }
        }

        printf("Generation %d: %s (Fitness: %d)\n", generation, population[0].genes, fitness_function(population[0]));

        // Create a new population
        Individual new_population[POPULATION_SIZE];
        for (int i = 0; i < POPULATION_SIZE; i++) {
            Individual parent1, parent2;
            select_parents(population, &parent1, &parent2);
            new_population[i] = crossover(parent1, parent2);
            mutate(&new_population[i]);
        }

        // Replace the old population with the new population
        memcpy(population, new_population, sizeof(population));

        // Check if target is found
        for (int i = 0; i < POPULATION_SIZE; i++) {
            if (strcmp(population[i].genes, TARGET) == 0) {
                printf("Target found in generation %d: %s\n", generation, TARGET);
                return;
            }
        }
    }

    printf("Target not found in %d generations.\n", GENERATIONS);
}